src/plugins/controls/common/options_list/ip_search.ts

```diff
@@ -0,0 +1,118 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+import ipaddr from 'ipaddr.js';
+
+export interface IpRangeQuery {
+  validSearch: boolean;
+  rangeQuery?: Array<{ key: string; from: string; to: string } | { key: string; mask: string }>;
+}
+interface IpSegments {
+  segments: string[];
+  type: 'ipv4' | 'ipv6' | 'unknown';
+}
+
+export const getIpSegments = (searchString: string): IpSegments => {
+  if (searchString.indexOf('.') !== -1) {
+    // ipv4 takes priority - so if search string contains both `.` and `:` then it will just be an invalid ipv4 search
+    const ipv4Segments = searchString.split('.').filter((segment) => segment !== '');
+    return { segments: ipv4Segments, type: 'ipv4' };
+  } else if (searchString.indexOf(':') !== -1) {
+    // note that currently, because of the logic of splitting here, searching for shorthand IPv6 IPs is not supported (for example,
+    // must search for `59fb:0:0:0:0:1005:cc57:6571` and not `59fb::1005:cc57:6571` to get the expected match)
+    const ipv6Segments = searchString.split(':').filter((segment) => segment !== '');
+    return { segments: ipv6Segments, type: 'ipv6' };
```

Comment:
Just drawing attention to this comment here, because I foresee that this may be a question that pops up eventually.

Basically, IPv4 supports shorthand where you are allowed a single `::` and this will basically say "pad here with zeros so that we have 8 segments". We could, in theory, support this - however, the logic gets kinda complicated. So, if users try to use this shorthand, they may get slightly unexpected results.

For example, suppose their data has the IP `59fb::1005:cc57:6571`. In order to find this as an exact match, they would have to search for `59fb:0:0:0:1005:cc57:6571`. Partial searching, on the other hand, would only be supported **up to** the `::` character - in this instance, a search of `59fb:` would return the expected IP but `59fb::1005` would **not**.